pulsar-broker/src/main/java/org/apache/pulsar/broker/service/PublishRateLimiterImpl.java

```diff
@@ -16,70 +16,133 @@
-    protected volatile boolean publishRateExceeded = false;
-    protected volatile LongAdder currentPublishMsgCount = new LongAdder();
-    protected volatile LongAdder currentPublishByteCount = new LongAdder();
+    private volatile AsyncTokenBucket tokenBucketOnMessage;
+    private volatile AsyncTokenBucket tokenBucketOnByte;
+    private final LongSupplier clockSource;
+
+    private final MessagePassingQueue<Producer> unthrottlingQueue = new MpscUnboundedArrayQueue<>(1024);
+
+    private final AtomicBoolean unthrottlingScheduled = new AtomicBoolean(false);
 
     public PublishRateLimiterImpl(Policies policies, String clusterName) {
+        this();
         update(policies, clusterName);
     }
 
     public PublishRateLimiterImpl(PublishRate maxPublishRate) {
+        this();
         update(maxPublishRate);
     }
 
+    public PublishRateLimiterImpl() {
+        this(AsyncTokenBucket.DEFAULT_CLOCK_SOURCE);
+    }
+
+    public PublishRateLimiterImpl(LongSupplier clockSource) {
+        this.clockSource = clockSource;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
     @Override
-    public void checkPublishRate() {
-        if (this.publishThrottlingEnabled && !publishRateExceeded) {
-            if (this.publishMaxByteRate > 0) {
-                long currentPublishByteRate = this.currentPublishByteCount.sum();
-                if (currentPublishByteRate > this.publishMaxByteRate) {
-                    publishRateExceeded = true;
-                    return;
-                }
-            }
-
-            if (this.publishMaxMessageRate > 0) {
-                long currentPublishMsgRate = this.currentPublishMsgCount.sum();
-                if (currentPublishMsgRate > this.publishMaxMessageRate) {
-                    publishRateExceeded = true;
-                }
-            }
+    public void handlePublishThrottling(Producer producer, int numOfMessages,
+                                        long msgSizeInBytes) {
+        boolean shouldThrottle = false;
+        AsyncTokenBucket currentTokenBucketOnMessage = tokenBucketOnMessage;
+        if (currentTokenBucketOnMessage != null) {
+            // consume tokens from the token bucket for messages
+            currentTokenBucketOnMessage.consumeTokens(numOfMessages);
+            // check if the token bucket contains remaining tokens, if not, we should throttle
+            shouldThrottle = !currentTokenBucketOnMessage.containsTokens();
+        }
+        AsyncTokenBucket currentTokenBucketOnByte = tokenBucketOnByte;
+        if (currentTokenBucketOnByte != null) {
+            // consume tokens from the token bucket for bytes
+            currentTokenBucketOnByte.consumeTokens(msgSizeInBytes);
+            // check if the token bucket contains remaining tokens, if not, we should throttle
+            shouldThrottle = shouldThrottle || !currentTokenBucketOnByte.containsTokens();
+        }
+        if (shouldThrottle) {
```

Comment:
The producer has no current knowledge of why it was rate limited (# msgs or # bytes). Do you envision adding context in a later PR (post V1)?

Reply:
yes, that's possible.